opentelemetry-exporter-geneva/geneva-uploader/src/config_service/client.rs

```diff
@@ -0,0 +1,536 @@
+                        }
+                    }
+                }
+            }
+
+            // Update cache with fresh data
+            cached_data.auth_info = Some((fresh_token_info, fresh_moniker_info));
+            cached_data.token_endpoint = Some(token_endpoint);
+            cached_data.token_expiry = token_expiry;
+
+            if let (Some(info), Some(endpoint)) =
+                (&cached_data.auth_info, &cached_data.token_endpoint)
+            {
+                Ok((info.0.clone(), info.1.clone(), endpoint.clone()))
+            } else {
+                Err(GenevaConfigClientError::InternalError(
+                    "Cache was not properly updated".to_string(),
+                ))
+            }
+        }
+    }
+
+    /// Internal method that actually fetches data from Geneva Config Service
+    async fn fetch_ingestion_info(&self) -> Result<(IngestionGatewayInfo, MonikerInfo)> {
+        let agent_identity = "GenevaUploader"; // TODO make this configurable
+        let agent_version = "0.1"; // TODO make this configurable
+        let identity = format!(
+            "Tenant=Default/Role=GcsClient/RoleInstance={}",
+            agent_identity
+        );
+
+        let encoded_identity = general_purpose::STANDARD.encode(&identity);
+        let version_str = format!("Ver{}v0", self.config.config_major_version);
+
+        // Construct the following URL:
+        // https://<endpoint>/api/agent/v3/<environment>/<account>/MonitoringStorageKeys/
+        //   ?Namespace=<namespace>
+        //   &Region=<region>
+        //   &Identity=<base64-encoded identity>
+        //   &OSType=<os_type>
+        //   &ConfigMajorVersion=Ver<major_version>v0
+        //   &TagId=<uuid>
+        let endpoint = self.config.endpoint.trim_end_matches('/');
+        let tag_id = Uuid::new_v4().to_string(); //TODO - uuid is costly, check if counter is enough?
+        let mut url = String::with_capacity(endpoint.len() + 200); // Pre-allocate with reasonable capacity
+        url.push_str(endpoint);
+        url.push_str("/api/agent/v3/");
+        url.push_str(&self.config.environment);
+        url.push('/');
+        url.push_str(&self.config.account);
+        url.push_str("/MonitoringStorageKeys/?Namespace=");
+        url.push_str(&self.config.namespace);
+        url.push_str("&Region=");
+        url.push_str(&self.config.region);
+        url.push_str("&Identity=");
+        url.push_str(&encoded_identity);
+        url.push_str("&OSType=");
+        url.push_str(get_os_type());
+        url.push_str("&ConfigMajorVersion=");
+        url.push_str(&version_str);
+        url.push_str("&TagId=");
+        url.push_str(&tag_id);
+
+        let req_id = Uuid::new_v4().to_string();
+
+        let response = self
+            .http_client
+            .get(&url)
+            .header(
+                "User-Agent",
+                format!("{}-{}", agent_identity, agent_version),
```

Comment:
nit: This header value doesn't change. We could avoid creating it for each request.

Reply:
Yes correct. Have precomputed and stored it with GenevaConfigClient now.